src/app/ryderserial/ryder-serial/index.ts

```diff
@@ -140,359 +131,345 @@ export default class RyderSerial extends Events.EventEmitter {
+        data: '0x' + Buffer.from(completed_data, 'binary').toString('hex'),
+        is_single_byte,
+      });
+
+      // Handle the completed response
+      this.on_device_response(completed_data, is_single_byte);
+    }
   }
 
-  /**
-   * Releases the last lock that was requested.
+  /*
+   * Handles completed responses that were reconstructed by `on_raw_device_response`.
    *
-   * Be sure to call this after calling `lock()`, otherwise the serial connection may be blocked until your
-   * app exits or the Ryder disconnects.
+   * @param response The complete response.
+   * @param is_single_byte Whether `response` represents a single-byte response from the device.
    */
-  public unlock(): void {
-    if (this[lock_symbol].length) {
-      this.log(LogLevel.DEBUG, 'ryderserial unlock');
-      const resolve = this[lock_symbol].shift();
-      resolve && resolve();
+  private on_device_response(response: Uint8Array, is_single_byte: boolean) {
+    // The response to return to the corresponding command if one is available, which may be either
+    // data or an error
+    let result;
+
+    if (is_single_byte === true) {
+      const byte = response[0];
+      switch (byte) {
+        // Simple responses that are just returned to the command
+        case RESPONSE_OK:
+        case RESPONSE_SEND_INPUT:
+        case RESPONSE_REJECTED:
+          result = response;
+          break;
+        // Special responses that are not returned to the command
+        case RESPONSE_WAIT_USER_CONFIRM:
+          // Update the state and notify the caller, but continue waiting for the followup response
+          this.set_state(State.WAITING_FOR_USER_CONFIRM);
+          this.emit('wait_user_confirm');
+          break;
+        case RESPONSE_LOCKED:
+          // The device should never be locked, so this response is an error
+          this.log(
+            LogLevel.ERROR,
+            'RESPONSE_LOCKED -- RYDER DEVICE IS NEVER SUPPOSED TO EMIT THIS EVENT'
+          );
+          this.emit('locked');
+          this.on_error(new Error('ERROR_LOCKED'));
+          break;
+        default:
+          if (byte in response_errors) {
+            // Error responses
+            result = new Error(response_errors[byte]);
+          } else {
+            // Unknown responses
+            result = new Error('ERROR_UNKNOWN_RESPONSE_' + byte);
+          }
+          break;
+      }
+    } else if (is_single_byte === false) {
+      // Multi-byte data is not special, so it is just returned
+      result = response;
+    }
+
+    // If a response to the command is available, return it and remove the command from the queue
+    if (result !== undefined) {
+      let { resolve, reject } = this.queue.remove();
+
+      if (result instanceof Error) {
+        reject(result);
+      } else {
+        // Convert the result to a string
+        let result_string = Buffer.from(result).toString('binary');
+        resolve(result_string);
```

Comment:
@MarvinJanssen This is now the only mention of strings when returning results to commands, so it will be easy to remove when the relevant changes are made to `ryder-utils`.